### ISSUE TITLE
Passing Gradle Proxy to non-compliant gradle tasks

### ISSUE THREAD

**Issue by pepper-jk** [oss-review-toolkit/ort#7359]
Hi ORT Team,

for the past couple of weeks we have been trying to set up IntelliJ on Windows Systems behind a Proxy.
We managed to set it up on our Linux machines, but ran into certain proxy issues on our windows machines.

When importing the ORT project IntelliJ runs a `gradle sync` of some sort.
This fails on our Windows machines, while it works fine on Linux.

We set the proxy settings in IntelliJ and in `gradle.properties` for both systems:
```
systemProp.http.proxyHost=proxy.com
systemProp.http.proxyPort=3128
systemProp.http.proxyUser=<username>
systemProp.http.proxyPassword=<password>
systemProp.http.nonProxyHosts=localhost|127.0.0.0/8

systemProp.https.proxyHost=proxy.com
systemProp.https.proxyPort=3128
systemProp.https.proxyUser=<username>
systemProp.https.proxyPassword=<password>
systemProp.http.nonProxyHosts=localhost|127.0.0.0/8
```

However, on Windows we encountered the following issue:
```
Could not create task ':utils:spdx-utils:generateSpdxEnums'.
Could not create task ':utils:spdx-utils:generateSpdxLicenseEnum'.
Unable to process url: https://scancode-licensedb.aboutcode.org/index.json
Unable to tunnel through proxy. Proxy returns "HTTP/1.1 407 AuthenticationRequired"
```

After some tinkering, I figured out a solution to "pass" the gradle proxy to the `groovy.json.JsonSlurper` like so:
```diff
diff --git a/utils/spdx/build.gradle.kts b/utils/spdx/build.gradle.kts
index 6bf9770cb..47856124f 100644
--- a/utils/spdx/build.gradle.kts
+++ b/utils/spdx/build.gradle.kts
@@ -21,6 +21,8 @@ import de.undercouch.gradle.tasks.download.Download

 import groovy.json.JsonSlurper

+import java.net.Authenticator
+import java.net.PasswordAuthentication
 import java.net.URL

 import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
@@ -74,6 +76,19 @@ class ScanCodeLicenseTextProvider : SpdxLicenseTextProvider {

         logger.quiet("Downloading ScanCode license index from $url...")

+        // Get Proxy login from gradle.properties
+        val proxyUser = project.ext["systemProp.https.proxyUser"].toString()
+        val proxyPassword = project.ext["systemProp.https.proxyPassword"].toString()
+
+        // Set proxy user and password as default authenticator
+        val authenticator: Authenticator = object : Authenticator() {
+            override fun getPasswordAuthentication() = PasswordAuthentication(
+                    proxyUser,
+                    proxyPassword.toCharArray()
+                )
+        }
+        Authenticator.setDefault(authenticator)
+
         val json = jsonSlurper.parse(url, "UTF-8")

         (json as List<Map<String, Any?>>).mapNotNull { map ->
```

After this change the **Sync** went through and IntelliJ listed the _tasks_ from the ORT project.

From what I understand the `java.net.Authenticator` is supposed to ask the user for credentials by default, but the gradle build environment suppresses that output and user intput, therefore the Java based tasks fail to use the Proxy.

Building works partially now. The main problems there occur with the `YarnInstall` task and the `Downloader` test, which use external package managers. Both fail with a `407` as they can not authenticate with the Proxy.

I was thinking there might be a way to apply this `Authenticator` fix to all gradle tasks and for the more specific one's like `yarn` and the external package managers to pass the proxy as well? Would that be something you could add as a feature "only requiring a `gradle.properties` proxy for development"?

If you got any other pointers on what is going wrong here, they would be appreciated.

**Comment by sschuberth** [comment]
> Windows Systems behind a Proxy.

Just curious, shouldn't our fellow Bosch users have the same issue on "Bosch-proxy" Windows machines? Any hints @mnonnenmacher, @MarcelBochtler, @nnobelis?

**Comment by mnonnenmacher** [comment]
> > Windows Systems behind a Proxy.
> 
> Just curious, shouldn't our fellow Bosch users have the same issue on "Bosch-proxy" Windows machines? Any hints @mnonnenmacher, @MarcelBochtler, @nnobelis?

I don't remember anyone mentioning this issue, but I'm also not sure if there is anyone left using Windows for development.

**Comment by sschuberth** [comment]
> After this change the Sync went through and IntelliJ listed the tasks from the ORT project.

Yes, generally that diff makes sense to me. I've created a slightly cleaner version at https://github.com/oss-review-toolkit/ort/pull/7426. @pepper-jk, mind testing whether that works to solve the sync issues as well?

> From what I understand the `java.net.Authenticator` is supposed to ask the user for credentials by default

I don't think so... AFAIK there's no default implementation of `Authenticator` (per OS), and the "Usually, it will do this by prompting the user for information" part of the docs in my understanding just refers to how implementations will usually get the credentials.

> The main problems there occur with the `YarnInstall` task and the `Downloader` test, which use external package managers. Both fail with a `407` as they can not authenticate with the Proxy.

Please try additionally setting the `http(s)_proxy` environment variables with authentication encoded into the proxy URL. Many external tools adhere to those variables.

> I was thinking there might be a way to apply this `Authenticator` fix to all gradle tasks

Gradle *tasks* (built-in and from plugins) already adhere to proxy settings. What's special here is that we use custom code that's not directly part of tasks; or put another way, the JSON download is not implemented as a Gradle task itself. That's why setting the `Authenticator` manually is required here.

**Comment by sschuberth** [comment]
@pepper-jk, so this has been resolved for the SPDX download tasks. Is there anything left to do here?

**Comment by pepper-jk** [comment]
Thanks for the mention. I will verify this tomorrow.
Sorry, I must have forgotten about this.

**Comment by pepper-jk** [comment]
I'm still having trouble with the `YarnInstall` Task during build. But that issue seems unrelated to our `JsonSlurper` solution from what you described. I figured as much, but wanted to be sure. Thanks for the feedback there. I'm sure it is just an environment variable issue for me now.

Since the issue with `gradle sync` and the `JsonSlurper` is resolved, I close the issue. Thanks for all the help.

If we come back to the Windows Development Environment in the future, we can always open a more specific one. For now we have decided to stick to a Linux Development Environment anyway.
